File: 2024/july/java/21.java
class Solution {
    private void isNull(int x , ArrayList<Integer> [] g){
        if (g[x] == null) {
            g[x] = new ArrayList<>();
        }
    }
    public int[][] buildMatrix(int k, int[][] row, int[][] col) {
        Deque<Integer> q = new ArrayDeque<>();
        // solve for rows
        ArrayList<Integer> [] g = new ArrayList[k+1];
        for (int i = 1; i <= k; ++i) isNull(i , g);
        int [] deg = new int [k+1];
        for (var con : row){
            deg[con[1]]++;
            g[con[0]].add(con[1]);

        }
        for (int i = 1; i <= k; ++i){
            if (deg[i] == 0){
                q.addFirst(i);
            }
        }
        var ansRows = new int[k+1];
        int idx = 0;
        while(!q.isEmpty()){
            int node = q.removeFirst();
            ansRows[node] = idx++;
            for (int ch : g[node]){
                if (--deg[ch] == 0){
                    q.addFirst(ch);
                }
            }
        }
        if (idx != k) {
            return new int[0][0];
        }
        idx = 0;
        q.clear();
        for (int i = 1; i <= k; ++i) {
            g[i].clear();
            deg[i] = 0;
        }
        // solve cols
        for (var con : col){
            deg[con[1]]++;
            g[con[0]].add(con[1]);
        }
        for (int i = 1; i <= k; ++i){
            if (deg[i] == 0){
                q.addFirst(i);
            }
        }
        var ans = new int[k][k];
         while(!q.isEmpty()){
            int node = q.removeFirst();
            ans[ansRows[node]][idx++] = node;
            for (int ch : g[node]){
                if (--deg[ch] == 0){
                    q.addFirst(ch);
                }
            }
        }
        if(idx != k) {
            return new int [0][0];
        }
        return ans;


    }
}

File: 2024/july/javaScript/21.js
/**
 * @param {number} k
 * @param {number[][]} rowConditions
 * @param {number[][]} colConditions
 * @return {number[][]}
 */
var buildMatrix = function(k, row, col) {
        let q = [];
        // solve for rows
        let g = new Array(k+1);
        for (let i = 1; i <= k; ++i) g[i] = [];
        let deg = new Array(k+1).fill(0);
        for (const [a , b] of row){
            deg[b]++;
            g[a].push(b);

        }
        for (let i = 1; i <= k; ++i){
            if (deg[i] === 0){
                q.push(i);
            }
        }
        let ansRows = new Array(k+1).fill(0);
        let idx = 0;
        while(q.length){
            const node = q.shift();
            ansRows[node] = idx++;
            g[node].forEach(ch => {
                if (--deg[ch] === 0) q.push(ch);
            });
        }
        if (idx != k) {
            return [];
        }
        idx = 0;
        q = [];
        for (let i = 1; i <= k; ++i) {
            g[i] = [];
            deg[i] = 0;
        }
        // solve cols
        for (const [a , b] of col){
            deg[b]++;
            g[a].push(b);
        }
        for (let i = 1; i <= k; ++i){
            if (deg[i] === 0){
               q.push(i);
            }
        }
        let ans = new Array(k);
        for (let i = 0; i <k; ++i) ans[i] = new Array(k).fill(0);
          while(q.length){
            const node = q.shift();
            ans[ansRows[node]][idx++] = node;
            g[node].forEach(ch => {
                if (--deg[ch] === 0) q.push(ch);
            });
        }
        if(idx != k) {
           return [];
        }
        return ans;
};
